### DISCUSSION TITLE
Trying to add new app but templates are a mess

### DISCUSSION THREAD

**Discussion by ronaldgreeff** [cookiecutter/cookiecutter-django#3843] [Q&A] [answered]
I've created a new app and added it to the project folder. Weirdly, generic views (ListView) expect templates to be in a different folder than normal views (View).

```
- requirements
- my_project
    - contrib
    ▒   - sites
    ▒       - migrations
    - static
    ▒   - css
    ▒   - fonts
    ▒   - images
    ▒   ▒   - favicons
    ▒   - js
    - templates
    ▒   - account
    ▒   - pages
    ▒   - my_app
    ▒           - item_list.html <- only works here (generic / ListView)
    ▒   - users
    - my_app
    ▒   - migrations
    ▒   - templates
    ▒           - item_detail.html <- only works here (View)
    - users
    ▒   - migrations
    ▒   - tests
    - utils
```

My settings:

```
ROOT_DIR = Path(__file__).resolve(strict=True).parent.parent.parent
APPS_DIR = ROOT_DIR / "my_project"

INSTALLED_APPS = [
    ...
    "my_project.users",
    "my_project.my_app",  #  also tried my_project.my_app.apps.MyAppConfig
    ...
]

TEMPLATES = [
    {
        "BACKEND": "django.template.backends.django.DjangoTemplates",
        "DIRS": [str(APPS_DIR / "templates")],
        "APP_DIRS": True,
        "OPTIONS": {
            "context_processors": [
                "django.template.context_processors.debug",
                "django.template.context_processors.request",
                "django.contrib.auth.context_processors.auth",
                "django.template.context_processors.i18n",
                "django.template.context_processors.media",
                "django.template.context_processors.static",
                "django.template.context_processors.tz",
                "django.contrib.messages.context_processors.messages",
                "my_project.users.context_processors.allauth_settings",
            ],
        },
    }
]
```

`my_project/my_app/apps.py`

```
from django.apps import AppConfig
from django.utils.translation import gettext_lazy as _


class MyAppConfig(AppConfig):
    # default_auto_field = "django.db.models.BigAutoField"
    name = "my_project.my_app"
    verbose_name = _("My_App") # not really this but you get it

    def ready(self):
        try:
            import my_project.my_app.signals  # noqa F401
        except ImportError:
            pass
```

Am I missing something?

**Comment by luzfcb** [comment, marked as answer]
I will try to answer the 3 questions that were asked:

### <a name="how-to-create-a-new-django-app-to-cookiecuter-django-based-project" href="#how-to-create-a-new-django-app-to-cookiecuter-django-based-project">1 - How to create a new django app to cookiecuter-django based project:</a>


1 - create the `<name-of-the-app>` app with python manage.py startapp
2 - move `<name-of-the-app>` directory to `<project_slug>` directory
3 - edit `<project_slug>/<name-of-the-app>/apps.py` and 
change `name = "<name-of-the-app>"` to `name = "<project_slug>.<name-of-the-app>"`
4 - add `"<project_slug>.<name-of-the-app>.apps.<NameOfTheAppConfigClass>",` on your [`LOCAL_APPS` on `config/settings/base.py`](https://github.com/cookiecutter/cookiecutter-django/blob/23ba1b129c1c5860f8aa61db2570f7e08ec2f351/%7B%7Bcookiecutter.project_slug%7D%7D/config/settings/base.py#L96)



### <a name="how-to-explicitly-customize-the-template-used-by-django-views-that-inherit-from-templateresponsemixin" href="#how-to-explicitly-customize-the-template-used-by-django-views-that-inherit-from-templateresponsemixin">2 - How to explicitly customize the template used by django views that inherit from _TemplateResponseMixin_:</a>

For the `ListView/DetailView` problem, it would probably be nice if you had it explicit and simply set `template_name`:

```
- requirements
- my_project
    - contrib
    ▒   - sites
    ▒       - migrations
    - static
    ▒   - css
    ▒   - fonts
    ▒   - images
    ▒   ▒   - favicons
    ▒   - js
    - templates
    ▒   - account
    ▒   - pages
    ▒   - my_app
    ▒           - item_list.html <- only works here (generic / ListView)
    ▒           - item_detail.html
    ▒   - users
    - my_app
    ▒   - migrations
    - users
    ▒   - migrations
    ▒   - tests
    - utils
```

```python

from django.views import generic
from my_project.my_app.models import MyModel


class MyModelBaseViewMixin:
    model = MyModel

class MyModelListView(MyModelBaseViewMixin, generic.ListView):
    template_name = 'my_app/item_list.html'

class MyModelDetailView(MyModelBaseViewMixin, generic.DetailView):
    template_name = 'my_app/item_detail.html'


```

`TemplateResponseMixin` docs: https://ccbv.co.uk/projects/Django/4.0/django.views.generic.base/TemplateResponseMixin





### <a name="how-to-customize-the-template-used-by-django-admin-views" href="#how-to-customize-the-template-used-by-django-admin-views">3 - How to customize the template used by django admin views</a>


if the problem is related to Django Admin integration, please customize the path to the correct template via the _Custom template options_ of Django Admin: https://docs.djangoproject.com/en/4.0/ref/contrib/admin/#custom-template-options
